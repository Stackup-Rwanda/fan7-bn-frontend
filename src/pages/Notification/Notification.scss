.notification {
  position: relative;
  width: 300px;
  background: #ffffff;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  font-family: 'Poppins', sans-serif;
  box-sizing: border-box;

  &__header {
    display: flex;
    flex-direction: row;
    align-items: center;
    padding: 1em 0.6em;
    border-bottom: 1px solid rgba(196, 196, 196, 0.3);

    span {
      margin: 0 auto;
      font-style: normal;
      font-weight: 500;
      font-size: 14px;
      color: rgba(0, 0, 0, 0.3) !important;
    }
  }

  &__list {
    list-style: none;
    display: block;
    // height: 100vh;
    // overflow-y: hidden;
    // scrollbar-width: thin;
    // scrollbar-color: #90a4ae #cfd8dc;

    // &::-webkit-scrollbar-track {
    //   -webkit-box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
    //   box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
    //   background-color: #cfd8dc;
    // }

    // &::-webkit-scrollbar {
    //   width: 6px;
    //   // background-color: #f5f5f5;
    // }

    // &::-webkit-scrollbar-thumb {
    //   background-color: #90a4ae;
    //   border-radius: 6px;
    //   border: 3px solid #cfd8dc;
    // }
  }

  &__item {
    display: block;
    font-style: normal;
    text-align: left;
    padding: 0 0.6em;
    color: #000000;
    border-bottom: 1px solid rgba(196, 196, 196, 0.3);

    &__title {
      display: flex;
      justify-content: space-between;
      align-items: center;
      margin: 0;
      padding: 0;
      font-size: 14px;
      line-height: 30px;

      img {
        width: 14px;
        color: rgba(0, 0, 0, 0.3) !important;
      }
    }

    &__content {
      font-size: 12px;
      font-weight: 300;
      line-height: 30px;
      color: #000000;
      text-decoration: none;

      &:hover,
      :active,
      :focus {
        color: #2681ff;
        text-decoration: underline;
      }
    }
  }

  &__footer {
    display: flex;
    justify-content: center;
    padding: 1em 0.5em;
    font-style: normal;
    font-size: 12px;

    &__button {
      color: #2681ff;

      &:hover,
      :active,
      :focus {
        text-decoration: underline;
      }
    }
  }

  button {
    display: block;
    font-weight: normal;
    text-align: center;
    white-space: nowrap;
    background-color: transparent;
    background-image: none;
    -ms-touch-action: manipulation;
    touch-action: manipulation;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    border: none;
    cursor: pointer;

    &:focus,
    &:active &:active:focus {
      outline: 0;
      background-image: none;
    }
  }
}
