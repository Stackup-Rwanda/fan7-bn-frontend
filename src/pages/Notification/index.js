import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { selectNotifications, selectLoading } from '../../store/modules/notification/selectors';
import history from '../../utils/helpers/history';
import {
  getNotifications,
  markOneAsRead,
  markAllAsRead,
} from '../../store/modules/notification/actions';
import settingsIcon from '../../assets/icons/settings-24px.svg';
import closeIcon from '../../assets/icons/close-24px.svg';
import './Notification.scss';

class Notification extends Component {
  constructor(props) {
    super(props);

    this.state = {};
    this.markAllAsRead = this.markAllAsRead.bind(this);
    this.handleRedirect = this.handleRedirect.bind(this);
  }

  markAllAsRead() {
    const { dispatch } = this.props;
    dispatch(markAllAsRead());
    dispatch(getNotifications());
  }

  handleRedirect(notificationId, requestId) {
    const { dispatch } = this.props;
    dispatch(markOneAsRead(notificationId));
    dispatch(getNotifications());
    history.push(`/requests/${requestId}`);
  }

  render() {
    const { notifications, loading } = this.props;

    return (
      <div className="notification">
        <div className="notification__header">
          <button type="button">
            <img src={settingsIcon} alt="settings" />
          </button>
          <span>Notifications</span>
        </div>
        <ul className="notification__list">
          {notifications.length > 0
            ? notifications.map((notification) => (
                <li key={notification.id} className="notification__item">
                  <h4 className="notification__item__title">{notification.type}</h4>
                  <a
                    className="notification__item__content"
                    // href="/request"
                    onClick={() => this.handleRedirect(notification.id, notification.request_id)}
                  >
                    {notification.message}
                  </a>
                </li>
              ))
            : null}
        </ul>
        <div className="notification__footer">
          {notifications.length > 0 ? (
            loading ? (
              <span>Marking all as read...</span>
            ) : (
              <button
                type="button"
                className="notification__footer__button"
                onClick={this.markAllAsRead}
              >
                Mark all as read
              </button>
            )
          ) : (
            <span>No unread notifications</span>
          )}
        </div>
      </div>
    );
  }
}

Notification.propTypes = {
  dispatch: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  notifications: PropTypes.arrayOf().isRequired,
};

const select = ({ notifications }) => ({
  notifications: selectNotifications(notifications),
  loading: selectLoading(notifications),
});

export default connect(select)(Notification);
