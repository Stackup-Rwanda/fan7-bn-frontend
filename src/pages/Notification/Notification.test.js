/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount, shallow } from 'enzyme';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { BrowserRouter as Router } from 'react-router-dom';
import Notification from './';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const history = {
  push: jest.fn(),
  replace: jest.fn(),
  createHref: jest.fn(),
};

describe('Notification Component Test', () => {
  it('should return true if component exists', () => {
    const initialState = {};
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <Notification />
        </Router>
      </Provider>
    );

    expect(wrapper.find('.notification').exists()).toBe(true);
  });

  it('should list notifications if found', () => {
    const initialState = {
      notifications: {
        notifications: [
          {
            id: 8,
            message: 'New request has been created, waiting for approval',
            request_id: 11,
            status: 'unread',
            type: 'Created',
          },
        ],
      },
    };
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <Notification />
        </Router>
      </Provider>
    );

    expect(wrapper.find('.notification__item').exists()).toBe(true);
  });

  it('should not list notifications if not found', () => {
    const initialState = {
      notifications: {
        notifications: [],
      },
    };
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <Notification />
        </Router>
      </Provider>
    );

    expect(wrapper.find('.notification__item').exists()).toBe(false);
  });

  it('should handle redirect', () => {
    const initialState = {
      notifications: {
        notifications: [
          {
            id: 8,
            message: 'New request has been created, waiting for approval',
            request_id: 11,
            status: 'unread',
            type: 'Created',
          },
        ],
      },
    };
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <Notification />
        </Router>
      </Provider>
    );

    wrapper.find('.notification__item__content').simulate('click');
    expect(wrapper.find('.notification__item__content').exists()).toBe(true);
  });

  it('should handle mark all notifications as read', () => {
    const initialState = {
      notifications: {
        notifications: [
          {
            id: 8,
            message: 'New request has been created, waiting for approval',
            request_id: 11,
            status: 'unread',
            type: 'Created',
          },
          {
            id: 9,
            message: 'New request has been created, waiting for approval',
            request_id: 12,
            status: 'unread',
            type: 'Created',
          },
        ],
      },
    };
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <Notification />
        </Router>
      </Provider>
    );

    wrapper.find('.notification__footer__button').simulate('click');
    expect(wrapper.find('.notification__footer__button').exists()).toBe(true);
  });

  it('should display loader while marking all notifications as read', () => {
    const initialState = {
      notifications: {
        notifications: [
          {
            id: 8,
            message: 'New request has been created, waiting for approval',
            request_id: 11,
            status: 'unread',
            type: 'Created',
          },
          {
            id: 9,
            message: 'New request has been created, waiting for approval',
            request_id: 12,
            status: 'unread',
            type: 'Created',
          },
        ],
        loading: true,
      },
    };
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <Notification />
        </Router>
      </Provider>
    );

    expect(wrapper.find('.notification__footer').text()).toEqual('Marking all as read...');
  });
});

describe('<Notification /> shallow rendering tests', () => {
  const initialState = {};
  const store = mockStore(initialState);

  it('matches the snapshot', () => {
    const tree = shallow(<Notification store={store} />);
    expect(tree).toMatchSnapshot();
  });
});
