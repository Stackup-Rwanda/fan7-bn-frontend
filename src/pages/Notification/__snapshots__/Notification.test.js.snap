// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notification /> shallow rendering tests matches the snapshot 1`] = `
<ContextProvider
  value={null}
>
  <Notification
    dispatch={[Function]}
    loading={false}
    notifications={Array []}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;
